src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -181,26 +184,35 @@ private bool Match(ServerBuildArtifact buildArtifact, ArtifactDefinition artifac
                 var fileShareArtifact = new FileShareArtifact();
                 await fileShareArtifact.DownloadArtifactAsync(executionContext, HostContext, artifactDefinition, fileShare, downloadFolderPath);
             }
-            else if (string.Equals(buildArtifact.Resource.Type, WellKnownArtifactResourceTypes.Container, StringComparison.OrdinalIgnoreCase))
+            else if (buildArtifactDetails != null &
+                     string.Equals(buildArtifact.Resource.Type, WellKnownArtifactResourceTypes.Container, StringComparison.OrdinalIgnoreCase))
             {
-                executionContext.Output("Artifact Type: ServerDrop");
-
-                // TODO:Get VssBinFetchclient and get away from zipstream downloader
-                Stream contentStream;
+                string[] parts = buildArtifact.Resource.Data.Split(new[] { '/' }, 3);
 
-                if (definitionType == DefinitionType.Xaml)
-                {
-                    contentStream = await xamlBuildClient.GetArtifactContentZipAsync(buildArtifactDetails.Project, buildId, buildArtifact.Name);
-                }
-                else
+                int containerId;
+                string rootLocation = parts[2];
+                if (!int.TryParse(parts[1], out containerId))
                 {
-                    contentStream = await buildClient.GetArtifactContentZipAsync(buildArtifactDetails.Project, buildId, buildArtifact.Name);
+                    throw new ArtifactDownloadException(string.Format(CultureInfo.CurrentCulture, "Resources.NotContainerType", buildArtifact.Name));
                 }
 
+                using (var engine = new ContainerFetchEngine(
+                    executionContext,
+                    buildArtifactDetails.TfsUrl,
+                    buildArtifactDetails.Project,
+                    containerId,
+                    rootLocation,
+                    Path.Combine(localFolderPath, rootLocation),
+                    parallelDownloadLimit: 4,
+                    verbose: false,
+                    accessToken: buildArtifactDetails.AccessToken))
+                {
+                    var fetchTask = engine.FetchAsync();
```

Comment:
It looks like `FetchAsync` is the method that ultimately calls ItemCache.GetItemsAsync(). Why is the underlying AsyncLazy and caching needed here? This uses a new ContainerFetchEngine every time and then disposes of it.